scarb/src/core/manifest/toml_manifest.rs

```diff
@@ -545,20 +548,32 @@ impl TomlManifest {
         kind: impl Into<SmolStr>,
         target: Option<&TomlTarget<T>>,
         default_name: &SmolStr,
-        default_source_path: &Utf8Path,
+        root: &Utf8Path,
     ) -> Result<Option<Target>> {
+        let default_source_path = root.join(DEFAULT_SOURCE_PATH);
         let Some(target) = target else {
             return Ok(None);
         };
 
+        let kind: SmolStr = kind.into();
+        if let Some(source_path) = &target.source_path {
+            if kind != Target::TEST && source_path.to_str().unwrap_or("") != DEFAULT_SOURCE_PATH {
+                bail!("`{kind}` target cannot define non default `source-path`");
```

Comment:
```suggestion
                bail!("`{kind}` target cannot specify custom `source-path`");
```

Removed double negation this way